Update libunwind to latest

janvorli opened the pull request:
This change updates libunwind copy by a fix for non-4kB page sizes
that is needed e.g. for Linux arm64 with kernel set to use 64kB page
sizes (RHEL8 is one such case).

It also applies a fix for incorrect register storage in ARM64
getcontext.

Close #68640

am11 commented on src/native/external/libunwind-version.txt:
```diff
@@ -1,10 +1,6 @@
-v1.6.2
-https://github.com/libunwind/libunwind/commit/b3ca1b59a795a617877c01fe5d299ab7a07ff29d
+main@e07b43c02d5cf1ea060c018fdf2e2ad34b7c7d80
+https://github.com/libunwind/libunwind/commit/e07b43c02d5cf1ea060c018fdf2e2ad34b7c7d80
```
This has pulled up a lot of unnecessary changes. An alternative fix is to keep pointing to released version 1.6.2, and apply patch from PR 330, as we have done for the entire loongarch64 support.

janvorli replied:
The changes that it has pulled  in all look ok and simple. Besides the loongarch64 changes that are now included in the main and the PR 330, there were only five PRs - one just a licence header update, one adding couple of ifdefs for freebsd, one fixing malloc issue, one fixing a race and one updating the latex doc.

am11 replied:
Loongarch and  other patches were applied on top of 1.6.2 without resetting to libunwind's master branch, which is still using version 1.6.1-rc1. My comment was we can apply PR 330 in similar way which fixes the issue.

If you prefer this approach, ok. I think it just makes the versioning a bit ambiguous, but not a big deal.

am11 commented:
> for Linux arm64 with kernel set to use 64kB page sizes (RHEL8 is one such case).

Apparently, R2R was also broken on such linux-arm64 system: https://github.com/dotnet/runtime/pull/68845#discussion_r864355088. We do not test against non-4K-page configuration in CI, so there could be more stuff broken besides libunwind.

janvorli commented on src/native/external/libunwind-version.txt:
```diff
@@ -1,10 +1,6 @@
 Reapply changes from https://github.com/dotnet/runtime/commit/1b5719c2e3dde393531eaeb5b5cde05dabeef5b8
-Apply https://github.com/libunwind/libunwind/pull/317
 Apply https://github.com/libunwind/libunwind/pull/333
 Apply https://github.com/libunwind/libunwind/pull/342
```
I was trying to be conservative and include only the merged in changes. The PR 353 is an obvious bug fix though, so I agree that it would make sense to apply that one too. 
The other two are related to walking over a signal frame and we never do that, we unwind that differently in PAL_VirtualUnwind.

janvorli commented:
> We do not test against non-4K-page configuration in CI, so there could be more stuff broken besides libunwind.

All coreclr pri1 tests are passing on my local system with this change.

am11 commented:
> The windows builds that pull in libunwind are failing as they don't know the sysconf that is now being called to get the page size. I need to fix it.

Ah probably need to add it `libunwind/src/win/pal-single-threaded.c` (we have `getpagesize` also defined there).

janvorli commented:
Sigh, there is another problem, this time due to the PR 327. The change has some invalid C standard code that Unix compilers compile because of an extended behavior, but VC does not:
```c
  const unsigned long full_path_buff_sz = 1024;
  char                full_path_buff[full_path_buff_sz];
```
`const` doesn't mean a constant, but rather a "read only" variable in C and VC doesn't support declaring arrays with non-constant size.

janvorli commented:
So I'll probably just follow your advice and pull in the two PRs (330 and 353)

janvorli committed:
Update libunwind

This change updates libunwind copy by a fix for non-4kB page sizes
that is needed e.g. for Linux arm64 with kernel set to use 64kB page
sizes (RHEL8 is one such case).

It also applies a fix for incorrect register storage in ARM64
getcontext.

janvorli commented:
@am11 I've changed the PR to just add the PR 330 and 353 and add the windows getconf implementation.

am11 reviewed (commented):
Thank you. I will submit your patch along with https://github.com/am11/libunwind/commit/c59e547193e1ab05360dcd1a1ae71c4b2c6e4845 upstream to make future updates easier.

Update: https://github.com/libunwind/libunwind/pull/358

janvorli commented:
The Windows arm legs are failing
```
2022-05-05T15:50:38.8303558Z init.c.obj : error LNK2019: unresolved external symbol _write@12 referenced in function _print_error@4
```
It seems the print error has pulled in "write" function that we also need to implement for Windows.

janvorli committed:
Add dummy `write` method implementation

am11 commented on src/native/external/libunwind-version.txt:
```diff
@@ -3,8 +3,10 @@ https://github.com/libunwind/libunwind/commit/b3ca1b59a795a617877c01fe5d299ab7a0
+Apply https://github.com/libunwind/libunwind/pull/330
 Apply https://github.com/libunwind/libunwind/pull/333
 Apply https://github.com/libunwind/libunwind/pull/342
+Apply https://github.com/libunwind/libunwind/pull/353
```
```suggestion
Apply https://github.com/libunwind/libunwind/pull/353
Apply https://github.com/libunwind/libunwind/pull/358
```

I have separated unrelated change in 359.

janvorli committed:
Update src/native/external/libunwind-version.txt

Co-authored-by: Adeel Mujahid <3840695+am11@users.noreply.github.com>

am11 reviewed (approved):
LGTM, thank you.
